spatialos/UnrealGDK
# Bugfix/unr 1285 character movement not replicated after possession

**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Character movements not replicated after possessing a character that was already possessed

Bug Flow:
1. Client 1 possess character actor.
2. Client 2 attempts to possess character actor.
3. Possession is processed on there server. 
4. The characterMovementComponent has it's CurrentClientTimeStamp reset to 0.
5. A couple of cheeky ServerMove RPCs are sent from Client 1 that are processed setting CurrentClientTimeStamp to a high value.
6. Client 2 now starts sending ServerMove RPCs but it's timestamp is lower than the CurrentClientTimeStamp.

The fix for the issue essentially stops these naughty RPCs getting through. This is achieved by:
1. The server asking for authority over the client RPC component.
2. Whilst this is being processed, any RPCs sent are ignored as we are "processing ownership transfer".
3. Once we gain authority over the component we transfer it's authority back to the requested client worker. 

#### Tests
Tested using packed and unpacked RPCs.

#### QA
1. Set up and environment in which you have multiple players and a possessable character actor. 
2. Possess the character actor with one of the players. Note that both players should be able to see the character actor move (movement is being replicated). 
3. Possess the character actor with the other player. Note that both players should be able to see the character actor move (movement is being replicated). 

If one of the players does not see the character moving the fix has failed. 

#### Primary reviewers
@joshuahuburn 
@improbable-valentyn
